In [1]:
import numpy as np 
import keras
from keras.models import Sequential
from keras.layers import Dense, Flatten, Dropout
from keras.optimizers import Adam
from keras.utils import np_utils 
from keras import backend as K 
import pandas as pd
from sklearn.model_selection import train_test_split

Using TensorFlow backend.


### Data set size
For this lab I used tha same data set that I produced in Lab 6, since it consists of 5000 observations (compared to example's 1000 observations), which has proven to impove accuracy.

In [43]:
data = pd.read_csv("dict1.csv", header = None, names = ["Angles", "XY"])

In [44]:
data.head(10)

Angles  \
0  (-0.40510000000000002, -0.097100000000000006, ...   
1  (-0.19320000000000001, 0.1341, -0.264699999999...   
2  (0.192, 0.155, -0.28570000000000001, -0.393199...   
3  (0.24679999999999999, -0.18140000000000001, 0....   
4  (-0.35449999999999998, 0.3926, 0.1613, 0.07829...   
5  (0.31819999999999998, -0.22770000000000001, -0...   
6  (0.4531, 0.085400000000000004, 0.4546, 0.09879...   
7  (0.1404, -0.065500000000000003, -0.08409999999...   
8  (0.34379999999999999, 0.27279999999999999, -0....   
9  (0.21460000000000001, -0.1293, -0.040399999999...   

                          XY  
0  [ 3.5048  0.9169  0.25  ]  
1  [ 3.6347  0.3976  0.25  ]  
2  [ 3.1567 -1.6357  0.25  ]  
3  [ 3.2989 -1.5492  0.25  ]  
4  [ 3.6155 -0.2022  0.25  ]  
5  [ 3.6511 -0.3356  0.25  ]  
6  [ 3.4324  0.4782  0.25  ]  
7  [ 3.6861  0.241   0.25  ]  
8  [ 3.1972 -1.6062  0.25  ]  
9  [ 3.6256  0.3925  0.25  ]

In [45]:
labels = data['Angles'].to_numpy()
train = data['XY'].to_numpy()

In [48]:
X = list()
Y = list()
for i in range(len(train)):
    train[i] = train[i].replace('   ', ' ')
    train[i] = train[i].replace('  ', ' ')
    train[i] = train[i].strip('[ ').strip(' ]')
    labels[i] = labels[i].strip('(').strip(')')
    result = [float(val) for val in labels[i].split(',')]
    Y.append(result)
    result = [float(val) for val in train[i].split(' ')]
    X.append(result)

In [49]:
X_train, X_test, y_train, y_test = train_test_split(np.asarray(X), np.asarray(Y), test_size=0.80)

In [50]:
print(np.shape(X_train))
print(np.shape(y_train))
print(np.shape(X_test))
print(np.shape(y_test))

(1000, 3)
(1000, 5)
(4001, 3)
(4001, 5)


### Loss function, layers and optimizer
In this lab I just used the most effective loss function from the Lab 6 (tested among MSLE, MAE, MSE and Regular RMSE). Besides, I addded one more hidden layer and used Adam optimizer, also based on the result obtained in Lab 6.

In [51]:
model = Sequential()
model.add(Dense(10, input_dim = 3, activation = 'relu'))
model.add(Dense(16, activation = 'relu'))
model.add(Dense(16, activation = 'relu'))
model.add(Dense(5, activation='linear'))
model.compile(loss='mean_squared_logarithmic_error', optimizer=Adam(0.01))

X_train, X_test, y_train, y_test = train_test_split(np.asarray(X), np.asarray(Y), test_size=0.80)
model.fit(X_train, y_train, epochs = 10, verbose = 0)
scores_msle = model.evaluate(X_test, y_test, verbose=0) 
print("RMSE: %.5f" % (scores_msle))

RMSE: 0.01806


In [52]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_19 (Dense)             (None, 10)                40        
_________________________________________________________________
dense_20 (Dense)             (None, 16)                176       
_________________________________________________________________
dense_21 (Dense)             (None, 16)                272       
_________________________________________________________________
dense_22 (Dense)             (None, 5)                 85        
Total params: 573
Trainable params: 573
Non-trainable params: 0
_________________________________________________________________


In [53]:
model.fit(X_train, y_train, epochs = 15)

Epoch 1/15
1000/1000 [==============================] - 0s 152us/step - loss: 0.0182
Epoch 2/15
1000/1000 [==============================] - 0s 122us/step - loss: 0.0183
Epoch 3/15
1000/1000 [==============================] - 0s 179us/step - loss: 0.0183
Epoch 4/15
1000/1000 [==============================] - 0s 152us/step - loss: 0.0183
Epoch 5/15
1000/1000 [==============================] - 0s 145us/step - loss: 0.0183
Epoch 6/15
1000/1000 [==============================] - 0s 137us/step - loss: 0.0183
Epoch 7/15
1000/1000 [==============================] - 0s 161us/step - loss: 0.0183
Epoch 8/15
1000/1000 [==============================] - 0s 150us/step - loss: 0.0183
Epoch 9/15
1000/1000 [==============================] - 0s 209us/step - loss: 0.0182
Epoch 10/15
1000/1000 [==============================] - 0s 134us/step - loss: 0.0183
Epoch 11/15
1000/1000 [==============================] - 0s 125us/step - loss: 0.0183
Epoch 12/15
1000/1000 [==============================] - 0s 146

In [55]:
scores = model.evaluate(X_test, y_test, verbose=0) 
print("RMSE: %.3f" % (scores))

RMSE: 0.018


In [56]:
print(model.predict(X_train[10].reshape(1,3)))
print(y_train[10])

[[ 0.1716738   0.22191347  0.19492155 -0.2310429   0.3181228 ]]
[0.2537 0.0664 0.0664 0.1766 0.4265]


# Final Performance:
### Old RMSE = 0.1 |     New RMSE = 0.018
### Used properties: 
Data-set of 5000 observations <br>
Loss function: MSLE - Mean Squared Logarithmic Error<br>
Number of layers: 4 <br>
Number of epochs: 15